sklearn/tree/_criterion.pyx

```diff
@@ -962,6 +965,289 @@ cdef class MSE(RegressionCriterion):
         impurity_left[0] /= self.n_outputs
         impurity_right[0] /= self.n_outputs
 
+cdef class MAE(RegressionCriterion):
+    """Mean absolute error impurity criterion"""
+    def __dealloc__(self):
+        """Destructor."""
+        free(self.node_medians)
+
+    cdef np.ndarray left_child
+    cdef np.ndarray right_child
+    cdef double* node_medians
+
+    def __cinit__(self, SIZE_t n_outputs):
+        """Initialize parameters for this criterion.
+
+        Parameters
+        ----------
+        n_outputs: SIZE_t
+            The number of targets to be predicted
+        """
+
+        # Default values
+        self.y = NULL
+        self.y_stride = 0
+        self.sample_weight = NULL
+
+        self.samples = NULL
+        self.start = 0
+        self.pos = 0
+        self.end = 0
+
+        self.n_outputs = n_outputs
+        self.n_node_samples = 0
+        self.weighted_n_node_samples = 0.0
+        self.weighted_n_left = 0.0
+        self.weighted_n_right = 0.0
+
+        # Allocate accumulators. Make sure they are NULL, not uninitialized,
+        # before an exception can be raised (which triggers __dealloc__).
+        self.node_medians = NULL
+
+        # Allocate memory for the accumulators
+        self.node_medians = <double*> calloc(n_outputs, sizeof(double))
+
+        if (self.node_medians == NULL):
+            raise MemoryError()
+
+        self.left_child = np.empty(n_outputs, dtype='object')
```

Comment:
Preallocating would be a good idea, since each of these are independent of each other. I don't think that creating an array of structs works in cython, at least it hasn't worked for me. If you know how to make it work, please let me know, it'd save me a lot of time!